[Flaking Test] UT TestWebSocketClient for closed connection

### Which jobs are failing?

https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-unit

### Which tests are failing?

kubernetes/staging/src/k8s.io/client-go/tools/remotecommand TestWebSocketClient_TextMessageTypeError 

### Since when has it been failing?

not very sure as I was noticing this while working on #121213 

### Testgrid link

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-unit

### Reason for failure (if possible)

```
(base) ➜  remotecommand git:(e2e_3085_kep) go test -v -run TestWebSocketClient_TextMessageTypeError -count 20
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.547910   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.549109   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.549676   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.552227   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.553229   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:56.549378   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59680->127.0.0.1:59679: use of closed network connection
E1014 19:57:56.549381   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59682->127.0.0.1:59681: use of closed network connection
E1014 19:57:56.549811   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59684->127.0.0.1:59683: use of closed network connection
E1014 19:57:56.552763   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59686->127.0.0.1:59685: use of closed network connection
E1014 19:57:56.553394   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59688->127.0.0.1:59687: use of closed network connection
E1014 19:58:01.562592   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59690->127.0.0.1:59689: write: broken pipe
    websocket_test.go:928: expect stream to be closed after connection is closed.
--- FAIL: TestWebSocketClient_TextMessageTypeError (30.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.560619   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.01s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.561513   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.562235   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.563078   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.563886   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.564797   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.565428   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.567187   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.568615   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.570444   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.01s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.581075   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.581634   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.582861   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.583384   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
FAIL
exit status 1
FAIL	k8s.io/client-go/tools/remotecommand	30.192s
```

### Anything else we need to know?

to reproduce:
cd kubernetes/staging/src/k8s.io/client-go/tools/remotecommand
go test -v -run TestWebSocketClient_TextMessageTypeError -count 50 ( or higher)

### Relevant SIG(s)

/sig api-machinery

possible hit of this issue https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121576/pull-kubernetes-unit/1718099199943249920
```
=== RUN   TestWebSocketClient_TextMessageTypeError
E1028 03:35:05.544138   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:35000->127.0.0.1:36191: use of closed network connection
E1028 03:35:06.524954   65205 conn.go:347] Unable to write frame to 0: io: read/write on closed pipe
◈�Z'7��y�}�ӌz��/��_�S���k�@vв���	g��.x'��q��y�'�)��#+H����`�����4YO��+(z��|��Ձ�^��������Y؆ZWu�ژ2��u��=�mu��@I���z�ެ�c�
E1028 03:35:06.525103   65205 conn.go:315] Error on socket receive: read tcp 127.0.0.1:36191->127.0.0.1:35000: use of closed network connection
E1028 03:35:10.518668   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:47448->127.0.0.1:35513: use of closed network connection
E1028 03:35:10.519776   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:47456->127.0.0.1:35513: use of closed network connection
E1028 03:35:10.520883   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:47458->127.0.0.1:35513: use of closed network connection
E1028 03:35:15.532625   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:52250->127.0.0.1:33223: write: broken pipe
    websocket_test.go:931: expect stream to be closed after connection is closed.
--- FAIL: TestWebSocketClient_TextMessageTypeError (30.00s)
```

> https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1736402057557970944
> 
>     * TestWebSocketClient_EmptyMessageHandled failed
> 
> 
> It should be the same.

We see this flake in our internal CI too. https://testgrid.k8s.io/ibm-unit-tests-ppc64le#Periodic%20unit%20test%20suite%20on%20ppc64le

Reproducible easily on local machine with stress tool.

```
[root@raji-workspace kubernetes]# /root/go/bin/stress ./remotecommand.test -test.run TestWebSocketClient_EmptyMessageHandled
5s: 32 runs so far, 0 failures
10s: 72 runs so far, 0 failures
15s: 106 runs so far, 0 failures
20s: 136 runs so far, 0 failures
25s: 160 runs so far, 0 failures
30s: 190 runs so far, 0 failures
35s: 220 runs so far, 0 failures
40s: 250 runs so far, 0 failures

/tmp/go-stress-20240220T235316-3632365858
E0220 23:53:36.973202  609878 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:37186->127.0.0.1:41569: write: broken pipe
--- FAIL: TestWebSocketClient_EmptyMessageHandled (30.01s)
    websocket_test.go:992: expect stream to be closed after connection is closed.
FAIL


ERROR: exit status 1


/tmp/go-stress-20240220T235316-1948136060
E0220 23:53:39.070138  609972 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:48842->127.0.0.1:45621: write: broken pipe
--- FAIL: TestWebSocketClient_EmptyMessageHandled (30.01s)
    websocket_test.go:992: expect stream to be closed after connection is closed.
FAIL


ERROR: exit status 1

45s: 288 runs so far, 2 failures (0.69%)
50s: 328 runs so far, 2 failures (0.61%)
55s: 364 runs so far, 2 failures (0.55%)
1m0s: 402 runs so far, 2 failures (0.50%)
1m5s: 440 runs so far, 2 failures (0.45%)
1m10s: 480 runs so far, 2 failures (0.42%)
1m15s: 520 runs so far, 2 failures (0.38%)
1m20s: 560 runs so far, 2 failures (0.36%)
1m25s: 596 runs so far, 2 failures (0.34%)
1m30s: 636 runs so far, 2 failures (0.31%)
1m35s: 676 runs so far, 2 failures (0.30%)
1m40s: 716 runs so far, 2 failures (0.28%)
1m45s: 754 runs so far, 2 failures (0.27%)
1m50s: 792 runs so far, 2 failures (0.25%)
1m55s: 832 runs so far, 2 failures (0.24%)
^C
[root@raji-workspace kubernetes]# uname -a
Linux raji-workspace.pokprv.stglabs.ibm.com 5.14.0-312.el9.ppc64le #1 SMP Thu May 11 16:13:37 UTC 2023 ppc64le ppc64le ppc64le GNU/Linux
[root@raji-workspace kubernetes]#
```

--- a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go
@@ -344,7 +344,7 @@ func (conn *Conn) handle(ws *websocket.Conn) {
 			continue
 		}
 		if _, err := conn.channels[channel].DataFromSocket(data); err != nil {
-			klog.Errorf("Unable to write frame to %d: %v\n%s", channel, err, string(data))
+			klog.Errorf("Unable to write frame (%d bytes) to %d: %v", len(data), channel, err)
 			continue
 		}
 	}

--- a/staging/src/k8s.io/client-go/tools/remotecommand/websocket.go
+++ b/staging/src/k8s.io/client-go/tools/remotecommand/websocket.go
@@ -187,6 +187,9 @@ type wsStreamCreator struct {
 	// map of stream id to stream; multiple streams read/write the connection
 	streams   map[byte]*stream
 	streamsMu sync.Mutex
+	// setStreamErr holds the error to return to anyone calling setStreams.
+	// this is populated in closeAllStreamReaders
+	setStreamErr error
 }
 
 func newWSStreamCreator(conn *gwebsocket.Conn) *wsStreamCreator {
@@ -202,10 +205,14 @@ func (c *wsStreamCreator) getStream(id byte) *stream {
 	return c.streams[id]
 }
 
-func (c *wsStreamCreator) setStream(id byte, s *stream) {
+func (c *wsStreamCreator) setStream(id byte, s *stream) error {
 	c.streamsMu.Lock()
 	defer c.streamsMu.Unlock()
+	if c.setStreamErr != nil {
+		return c.setStreamErr
+	}
 	c.streams[id] = s
+	return nil
 }
 
 // CreateStream uses id from passed headers to create a stream over "c.conn" connection.
@@ -228,7 +235,11 @@ func (c *wsStreamCreator) CreateStream(headers http.Header) (httpstream.Stream, 
 		connWriteLock: &c.connWriteLock,
 		id:            id,
 	}
-	c.setStream(id, s)
+	if err := c.setStream(id, s); err != nil {
+		_ = s.writePipe.Close()
+		_ = s.readPipe.Close()
+		return nil, err
+	}
 	return s, nil
 }
 
@@ -312,13 +323,19 @@ func (c *wsStreamCreator) readDemuxLoop(bufferSize int, period time.Duration, de
 }
 
 // closeAllStreamReaders closes readers in all streams.
-// This unblocks all stream.Read() calls.
+// This unblocks all stream.Read() calls, and keeps any future streams from being created.
 func (c *wsStreamCreator) closeAllStreamReaders(err error) {
 	c.streamsMu.Lock()
 	defer c.streamsMu.Unlock()
 	for _, s := range c.streams {
 		// Closing writePipe unblocks all readPipe.Read() callers and prevents any future writes.
 		_ = s.writePipe.CloseWithError(err)
+	}
+	// ensure callers to setStreams receive an error after this point
+	if err != nil {
+		c.setStreamErr = err
+	} else {
+		c.setStreamErr = fmt.Errorf("closed all streams")
 	}
 }
 

--- a/staging/src/k8s.io/client-go/tools/remotecommand/websocket_test.go
+++ b/staging/src/k8s.io/client-go/tools/remotecommand/websocket_test.go
@@ -1116,6 +1116,14 @@ func TestWebSocketClient_HeartbeatSucceeds(t *testing.T) {
 	wg.Wait()
 }
 
+func TestLateStreamCreation(t *testing.T) {
+	c := newWSStreamCreator(nil)
+	c.closeAllStreamReaders(nil)
+	if err := c.setStream(0, nil); err == nil {
+		t.Fatal("expected error adding stream after closeAllStreamReaders")
+	}
+}
+
 func TestWebSocketClient_StreamsAndExpectedErrors(t *testing.T) {
 	// Validate Stream functions.
 	c := newWSStreamCreator(nil)

--- a/staging/src/k8s.io/client-go/tools/remotecommand/websocket_test.go
+++ b/staging/src/k8s.io/client-go/tools/remotecommand/websocket_test.go
@@ -817,6 +817,8 @@ func TestWebSocketClient_BadHandshake(t *testing.T) {
 // TestWebSocketClient_HeartbeatTimeout tests the heartbeat by forcing a
 // timeout by setting the ping period greater than the deadline.
 func TestWebSocketClient_HeartbeatTimeout(t *testing.T) {
+	blockRequestCtx, unblockRequest := context.WithCancel(context.Background())
+	defer unblockRequest()
 	// Create fake WebSocket server which blocks.
 	websocketServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 		conns, err := webSocketServerStreams(req, w, streamOptionsFromRequest(req))
@@ -824,8 +826,7 @@ func TestWebSocketClient_HeartbeatTimeout(t *testing.T) {
 			t.Fatalf("error on webSocketServerStreams: %v", err)
 		}
 		defer conns.conn.Close()
-		// Block server; heartbeat timeout (or test timeout) will fire before this returns.
-		time.Sleep(1 * time.Second)
+		<-blockRequestCtx.Done()
 	}))
 	defer websocketServer.Close()
 	// Create websocket client connecting to fake server.
@@ -840,8 +841,8 @@ func TestWebSocketClient_HeartbeatTimeout(t *testing.T) {
 	}
 	streamExec := exec.(*wsStreamExecutor)
 	// Ping period is greater than the ping deadline, forcing the timeout to fire.
-	pingPeriod := 20 * time.Millisecond
-	pingDeadline := 5 * time.Millisecond
+	pingPeriod := wait.ForeverTestTimeout // this lets the heartbeat deadline expire without renewing it
+	pingDeadline := time.Second           // this gives setup 1 second to establish streams
 	streamExec.heartbeatPeriod = pingPeriod
 	streamExec.heartbeatDeadline = pingDeadline
 	// Send some random data to the websocket server through STDIN.
@@ -859,8 +860,7 @@ func TestWebSocketClient_HeartbeatTimeout(t *testing.T) {
 	}()
 
 	select {
-	case <-time.After(pingPeriod * 5):
-		// Give up after about five ping attempts
+	case <-time.After(wait.ForeverTestTimeout):
 		t.Fatalf("expected heartbeat timeout, got none.")
 	case err := <-errorChan:
 		// Expecting heartbeat timeout error.

